File: src/test/java/nz/co/searchwellington/filters/attributesetters/FeedAttributeSetterTest.java
package nz.co.searchwellington.filters.attributesetters;

import static org.junit.Assert.assertEquals;
import static org.mockito.Mockito.when;
import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.repositories.HibernateResourceDAO;

import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.springframework.mock.web.MockHttpServletRequest;

public class FeedAttributeSetterTest {
	
	@Mock HibernateResourceDAO resourceDAO;
	@Mock Feed feed;
	private MockHttpServletRequest request;
	private FeedAttributeSetter feedAttributeSetter;
	
	@Before
	public void setup() {
		MockitoAnnotations.initMocks(this);
		request = new MockHttpServletRequest();
		when(feed.getName()).thenReturn("Wellington City Council news");
		when(resourceDAO.loadFeedByUrlWords("wcc-news")).thenReturn(feed);
		feedAttributeSetter = new FeedAttributeSetter(resourceDAO);		
	}
	
	@Test
	public void shouldSetFeedAttributeForFeedPagePath() {
		request.setPathInfo("/feed/wcc-news");		
		feedAttributeSetter.setAttributes(request);		
		assertEquals(feed, request.getAttribute(FeedAttributeSetter.FEED_ATTRIBUTE));		
	}
	
	@Test
	public void shouldSetFeedAttributeForFeedEditPagePath() {
		request.setPathInfo("/feed/wcc-news/edit");
		feedAttributeSetter.setAttributes(request);
		assertEquals(feed, request.getAttribute(FeedAttributeSetter.FEED_ATTRIBUTE));		
	}
	
	@Test
	public void shouldSetFeedAttributeForFeedSavePath() {
		request.setPathInfo("/feed/wcc-news/save");
		feedAttributeSetter.setAttributes(request);
		assertEquals(feed, request.getAttribute(FeedAttributeSetter.FEED_ATTRIBUTE));		
	}
	
}

File: src/test/scala/nz/co/searchwellington/filters/attributesetters/FeedAttributeSetterTest.scala
package nz.co.searchwellington.filters.attributesetters

import nz.co.searchwellington.ReasonableWaits
import nz.co.searchwellington.model.Feed
import nz.co.searchwellington.repositories.mongo.MongoRepository
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeEach
import org.mockito.Mockito.{mock, when}
import org.springframework.mock.web.MockHttpServletRequest

import java.util.UUID
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Await, Future}

class FeedAttributeSetterTest extends ReasonableWaits{
  private val mongoRepository = mock(classOf[MongoRepository])
  private val feed = Feed(id = UUID.randomUUID().toString, title = "Wellington City Council news")
  private val request = new MockHttpServletRequest

  private val feedAttributeSetter = new FeedAttributeSetter(mongoRepository)

  @BeforeEach
  def setup(): Unit = {
    when(mongoRepository.getFeedByUrlwords("wcc-news")).thenReturn(Future.successful(Some(feed)))
  }

  @Test def shouldSetFeedAttributeForFeedPagePath(): Unit = {
    request.setRequestURI("/feed/wcc-news")

    val attributes = Await.result(feedAttributeSetter.setAttributes(request), TenSeconds)

    assertEquals(feed, attributes("feedAttribute"))
  }

  @Test def shouldSetFeedAttributeForFeedEditPagePath(): Unit = {
    request.setRequestURI("/feed/wcc-news/edit")

    val attributes = Await.result(feedAttributeSetter.setAttributes(request), TenSeconds)

    assertEquals(feed, attributes("feedAttribute"))
  }

  @Test def shouldSetFeedAttributeForFeedSavePath(): Unit = {
    request.setRequestURI("/feed/wcc-news/save")

    val attributes = Await.result(feedAttributeSetter.setAttributes(request), TenSeconds)

    assertEquals(feed, attributes("feedAttribute"))
  }

  @Test def shouldSetFeedAttributeForAcceptAll(): Unit = {
    request.setRequestURI("/feed/wcc-news/accept-all")

    val attributes = Await.result(feedAttributeSetter.setAttributes(request), TenSeconds)

    assertEquals(feed, attributes("feedAttribute"))
  }

}
